rclpy/rclpy/executors.py

```diff
@@ -440,14 +460,15 @@ def _wait_for_ready_callbacks(
         :param timeout_sec: Seconds to wait. Block forever if ``None`` or negative.
             Don't wait if 0.
         :param nodes: A list of nodes to wait on. Wait on all nodes if ``None``.
+        :param condition: A callable that makes the function return immedaitely when it's True.
         """
         timeout_timer = None
         timeout_nsec = timeout_sec_to_nsec(timeout_sec)
         if timeout_nsec > 0:
             timeout_timer = Timer(None, None, timeout_nsec, self._clock, context=self._context)
 
         yielded_work = False
-        while not yielded_work and not self._is_shutdown:
+        while not yielded_work and not self._is_shutdown and not condition():
```

Comment:
I guess this solves the problem. If the subscription callback were being executed in another thread, and that thread calls `future.done()` just after this loop is entered, it would still be ok because the callback would trigger the executor's guard condition after completing and that would cause this loop to re-run.

Reply:
> I guess this solves the problem. If the subscription callback were being executed in another thread, and that thread calls `future.done()` just after this loop is entered, it would still be ok because the callback would trigger the executor's guard condition after completing and that would cause this loop to re-run.

That was the idea.
At some point I deleted the callback, I've re-added it in https://github.com/ros2/rclpy/pull/605/commits/7864a7b9fd1f8f63ff5d2c623a774246bbfb54b7.